File: backend/app/providers.py
"""Modular search providers. Credentials stay on the server."""
from __future__ import annotations

import os
from typing import Any

import httpx


class SearchError(Exception):
    def __init__(self, message: str, retryable: bool = True):
        super().__init__(message)
        self.retryable = retryable


class BaseProvider:
    name = "base"

    def configured(self) -> bool:
        return False

    def search(self, query: str, num: int = 10) -> list[dict[str, Any]]:
        raise NotImplementedError


class SerpApiProvider(BaseProvider):
    name = "serpapi"

    def configured(self) -> bool:
        return bool(os.getenv("SEARCH_API_KEY"))

    def search(self, query: str, num: int = 10) -> list[dict[str, Any]]:
        key = os.getenv("SEARCH_API_KEY")
        if not key:
            raise SearchError("SEARCH_API_KEY is not set for SerpAPI", retryable=False)
        params = {
            "engine": "google",
            "q": query,
            "api_key": key,
            "num": min(num, 20),
            "gl": "za",
            "hl": "en",
        }
        with httpx.Client(timeout=30) as client:
            r = client.get("https://serpapi.com/search.json", params=params)
        if r.status_code == 429:
            raise SearchError("SerpAPI rate limit", retryable=True)
        if r.status_code >= 400:
            raise SearchError(f"SerpAPI error {r.status_code}: {r.text[:300]}", retryable=r.status_code >= 500)
        data = r.json()
        if data.get("error"):
            raise SearchError(str(data["error"]), retryable=True)
        out = []
        for i, item in enumerate(data.get("organic_results") or [], start=1):
            out.append({
                "title": item.get("title") or "",
                "url": item.get("link") or "",
                "snippet": item.get("snippet") or "",
                "position": item.get("position") or i,
            })
        return out


class GoogleCseProvider(BaseProvider):
    name = "google_cse"

    def configured(self) -> bool:
        return bool(os.getenv("SEARCH_API_KEY") and os.getenv("GOOGLE_CSE_ID"))

    def search(self, query: str, num: int = 10) -> list[dict[str, Any]]:
        key = os.getenv("SEARCH_API_KEY")
        cx = os.getenv("GOOGLE_CSE_ID")
        if not key or not cx:
            raise SearchError("SEARCH_API_KEY and GOOGLE_CSE_ID required", retryable=False)
        params = {"key": key, "cx": cx, "q": query, "num": min(num, 10), "gl": "za"}
        with httpx.Client(timeout=30) as client:
            r = client.get("https://www.googleapis.com/customsearch/v1", params=params)
        if r.status_code == 429:
            raise SearchError("Google CSE rate limit", retryable=True)
        if r.status_code >= 400:
            raise SearchError(f"Google CSE error {r.status_code}: {r.text[:300]}", retryable=r.status_code >= 500)
        data = r.json()
        out = []
        for i, item in enumerate(data.get("items") or [], start=1):
            out.append({
                "title": item.get("title") or "",
                "url": item.get("link") or "",
                "snippet": item.get("snippet") or "",
                "position": i,
            })
        return out


class BraveProvider(BaseProvider):
    name = "brave"

    def configured(self) -> bool:
        return bool(os.getenv("BRAVE_API_KEY") or os.getenv("SEARCH_API_KEY"))

    def search(self, query: str, num: int = 10) -> list[dict[str, Any]]:
        key = os.getenv("BRAVE_API_KEY") or os.getenv("SEARCH_API_KEY")
        if not key:
            raise SearchError("BRAVE_API_KEY is not set", retryable=False)
        headers = {"Accept": "application/json", "X-Subscription-Token": key}
        params = {"q": query, "count": min(num, 20), "country": "ZA"}
        with httpx.Client(timeout=30) as client:
            r = client.get("https://api.search.brave.com/res/v1/web/search", params=params, headers=headers)
        if r.status_code == 429:
            raise SearchError("Brave rate limit", retryable=True)
        if r.status_code >= 400:
            raise SearchError(f"Brave error {r.status_code}: {r.text[:300]}", retryable=r.status_code >= 500)
        data = r.json()
        out = []
        for i, item in enumerate((data.get("web") or {}).get("results") or [], start=1):
            out.append({
                "title": item.get("title") or "",
                "url": item.get("url") or "",
                "snippet": item.get("description") or "",
                "position": i,
            })
        return out


class BingProvider(BaseProvider):
    name = "bing"

    def configured(self) -> bool:
        return bool(os.getenv("BING_API_KEY") or os.getenv("SEARCH_API_KEY"))

    def search(self, query: str, num: int = 10) -> list[dict[str, Any]]:
        key = os.getenv("BING_API_KEY") or os.getenv("SEARCH_API_KEY")
        if not key:
            raise SearchError("BING_API_KEY is not set", retryable=False)
        headers = {"Ocp-Apim-Subscription-Key": key}
        params = {"q": query, "count": min(num, 50), "mkt": "en-ZA"}
        with httpx.Client(timeout=30) as client:
            r = client.get("https://api.bing.microsoft.com/v7.0/search", params=params, headers=headers)
        if r.status_code == 429:
            raise SearchError("Bing rate limit", retryable=True)
        if r.status_code >= 400:
            raise SearchError(f"Bing error {r.status_code}: {r.text[:300]}", retryable=r.status_code >= 500)
        data = r.json()
        out = []
        for i, item in enumerate((data.get("webPages") or {}).get("value") or [], start=1):
            out.append({
                "title": item.get("name") or "",
                "url": item.get("url") or "",
                "snippet": item.get("snippet") or "",
                "position": i,
            })
        return out


PROVIDERS = {
    "serpapi": SerpApiProvider,
    "google_cse": GoogleCseProvider,
    "brave": BraveProvider,
    "bing": BingProvider,
}


def get_provider(name: str | None = None) -> BaseProvider:
    name = (name or os.getenv("SEARCH_PROVIDER") or "serpapi").lower()
    cls = PROVIDERS.get(name)
    if not cls:
        raise SearchError(f"Unknown provider {name}", retryable=False)
    return cls()


def provider_status() -> dict:
    current = (os.getenv("SEARCH_PROVIDER") or "serpapi").lower()
    return {
        "current": current,
        "providers": {
            k: {"configured": cls().configured()}
            for k, cls in PROVIDERS.items()
        },
    }

File: lib/providers.js
function env(name) {
  return process.env[name] || "";
}

const RETRYABLE_STATUSES = new Set([408, 429]);

function compact(value) {
  return String(value || "").replace(/\s+/g, " ").trim().slice(0, 300);
}

function errorDetails(payload, raw) {
  const error = payload && payload.error;
  if (typeof error === "string") return error;
  if (error && typeof error === "object") {
    const reasons = Array.isArray(error.errors)
      ? error.errors.map((item) => item && item.reason).filter(Boolean)
      : [];
    const status = error.status;
    const message = error.message;
    return [...new Set([...reasons, status, message].map(compact).filter(Boolean))].join(" — ");
  }
  return compact(payload?.message || raw);
}

function providerError(name, status, payload, raw = "") {
  const detail = errorDetails(payload, raw);
  const message = `${name} error ${status}${detail ? `: ${detail}` : ""}`;
  return Object.assign(new Error(message), {
    provider: name,
    status,
    reason: detail,
    retryable: RETRYABLE_STATUSES.has(status) || status >= 500,
  });
}

async function responseError(name, response) {
  let raw = "";
  try { raw = await response.text(); } catch {}
  let payload = null;
  try { payload = raw ? JSON.parse(raw) : null; } catch {}
  return providerError(name, response.status, payload, raw);
}

async function responseJson(name, response) {
  if (!response.ok) throw await responseError(name, response);
  try {
    return await response.json();
  } catch {
    throw Object.assign(new Error(`${name} returned an invalid JSON response`), {
      provider: name,
      retryable: true,
    });
  }
}

export function providerStatus() {
  const current = (env("SEARCH_PROVIDER") || "serpapi").toLowerCase();
  return {
    current,
    providers: {
      serpapi: { configured: Boolean(env("SEARCH_API_KEY")) },
      google_cse: { configured: Boolean(env("SEARCH_API_KEY") && env("GOOGLE_CSE_ID")) },
      brave: { configured: Boolean(env("BRAVE_API_KEY") || env("SEARCH_API_KEY")) },
      bing: { configured: Boolean(env("BING_API_KEY") || env("SEARCH_API_KEY")) },
    },
  };
}

export async function runSearch(query, num = 10) {
  const name = (env("SEARCH_PROVIDER") || "serpapi").toLowerCase();
  if (name === "google_cse") return googleCse(query, num);
  if (name === "brave") return brave(query, num);
  if (name === "bing") return bing(query, num);
  if (name === "serpapi") return serpapi(query, num);
  throw Object.assign(new Error(`Unknown search provider "${name}"`), { retryable: false });
}

async function serpapi(query, num) {
  const key = env("SEARCH_API_KEY");
  if (!key) throw Object.assign(new Error("SEARCH_API_KEY is not set for SerpAPI"), { retryable: false });
  const u = new URL("https://serpapi.com/search.json");
  u.searchParams.set("engine", "google");
  u.searchParams.set("q", query);
  u.searchParams.set("api_key", key);
  u.searchParams.set("num", String(Math.min(num, 20)));
  u.searchParams.set("gl", "za");
  u.searchParams.set("hl", "en");
  const r = await fetch(u);
  const data = await responseJson("SerpAPI", r);
  if (data.error) throw providerError("SerpAPI", 400, data);
  return (data.organic_results || []).map((item, i) => ({
    title: item.title || "",
    url: item.link || "",
    snippet: item.snippet || "",
    position: item.position || i + 1,
  }));
}

async function googleCse(query, num) {
  const key = env("SEARCH_API_KEY");
  const cx = env("GOOGLE_CSE_ID");
  if (!key || !cx) throw Object.assign(new Error("SEARCH_API_KEY and GOOGLE_CSE_ID required"), { retryable: false });
  const u = new URL("https://www.googleapis.com/customsearch/v1");
  u.searchParams.set("key", key);
  u.searchParams.set("cx", cx);
  u.searchParams.set("q", query);
  u.searchParams.set("num", String(Math.min(num, 10)));
  u.searchParams.set("gl", "za");
  const r = await fetch(u);
  const data = await responseJson("Google CSE", r);
  return (data.items || []).map((item, i) => ({
    title: item.title || "",
    url: item.link || "",
    snippet: item.snippet || "",
    position: i + 1,
  }));
}

async function brave(query, num) {
  const key = env("BRAVE_API_KEY") || env("SEARCH_API_KEY");
  if (!key) throw Object.assign(new Error("BRAVE_API_KEY is not set"), { retryable: false });
  const u = new URL("https://api.search.brave.com/res/v1/web/search");
  u.searchParams.set("q", query);
  u.searchParams.set("count", String(Math.min(num, 20)));
  u.searchParams.set("country", "ZA");
  const r = await fetch(u, { headers: { Accept: "application/json", "X-Subscription-Token": key } });
  const data = await responseJson("Brave", r);
  return ((data.web && data.web.results) || []).map((item, i) => ({
    title: item.title || "",
    url: item.url || "",
    snippet: item.description || "",
    position: i + 1,
  }));
}

async function bing(query, num) {
  const key = env("BING_API_KEY") || env("SEARCH_API_KEY");
  if (!key) throw Object.assign(new Error("BING_API_KEY is not set"), { retryable: false });
  const u = new URL("https://api.bing.microsoft.com/v7.0/search");
  u.searchParams.set("q", query);
  u.searchParams.set("count", String(Math.min(num, 50)));
  u.searchParams.set("mkt", "en-ZA");
  const r = await fetch(u, { headers: { "Ocp-Apim-Subscription-Key": key } });
  const data = await responseJson("Bing", r);
  return ((data.webPages && data.webPages.value) || []).map((item, i) => ({
    title: item.name || "",
    url: item.url || "",
    snippet: item.snippet || "",
    position: i + 1,
  }));
}
